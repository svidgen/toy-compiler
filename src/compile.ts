import process from 'process';
import { readFileSync } from 'fs';
import MonkeyScript from './monkeyscript-lang';

const ast = MonkeyScript.parse({
	code: readFileSync(process.argv[2], 'utf-8').trim()
});

console.dir(ast, { depth: 2 });
