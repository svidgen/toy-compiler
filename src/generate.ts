import { AST } from './parser';

/**
 * The output.
 */

const output = (ast: AST): string => {
    switch (ast.type) {
        case 'GAME':
            return null;
        case 'STATEMENT':
        case 'STATEMENT_BODY':
            return ast.children
                .map(c => outputAsyncJS(c))
                .filter(o => o).join(";")
            ;
        case 'ASSIGNMENT':
            return ast.children
                .filter(c => c.type !== 'WHITESPACE')
                .map(c => outputAsyncJS(c)).join(" ")
            ;
        case 'ASSIGNMENT_OPERATOR':
            return '=';
        case 'NAME':
            return ast.code;
        case 'VALUE':
            return 'await ' + ast.code;
        case 'EOS':
        case 'WHITESPACE':
        case 'NOISE':
            return '';
        default:
            throw new Error(`Unknown token "${ast.type}":\n${ast.code}`)
    }
};

/**
 * Execution.
 */
const grammar = new Sequence('PROGRAM', STATEMENT);
const syntaxTree = grammar.parse({code: `
    xyz <- 123 ;;;

    ;  abc <- xyz
`});
const js = syntaxTree ? outputAsyncJS(syntaxTree) : '';

console.clear();
console.log({
    syntaxTree,
    js
});
