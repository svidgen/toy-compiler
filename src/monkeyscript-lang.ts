import {
	TreePattern,
	AST,
	Sequence,
	Union,
	Recipe,
	Token
} from './parser';


/**
 * The grammar.
 */
const EOS = new Token("EOS", /\s*[\n|;]+\s*/);
const ASSIGNMENT_OPERATOR = new Token("ASSIGNMENT_OPERATOR", /^=/);
const WHITESPACE = new Token("WHITESPACE", /^\s+/, true);
const NAME = new Token("NAME", /^[a-zA-Z_][a-zA-Z0-9_]*/);
const OPERATOR = new Token("OPERATOR", /^[+*]/);
const DEF_SEPARATOR = new Token("DEF_SEPARATOR", /^:/);
const INTEGER = new Token("INTEGER", /^[0-9]+/);
const FLOAT = new Token("FLOAT", /^[0-9]*\.[0-9]+/);
const NUMBER = new Union("NUMBER", [ INTEGER, FLOAT ]);
const VALUE = new Union("VALUE", [ NAME, NUMBER ]);

const EXPRESSION = new Union(
	"EXPRESSION",
	[ "OPERATION", NAME, VALUE ]
);

const OPERAND = new Union(
	"OPERAND",
	[ NAME, VALUE ]
);

const OPERATION = new Recipe(
	"OPERATION",
	[ OPERAND, WHITESPACE, OPERATOR, WHITESPACE, OPERAND ]
);

const ASSIGNMENT = new Recipe(
    "ASSIGNMENT",
    [ NAME, WHITESPACE, ASSIGNMENT_OPERATOR, WHITESPACE, EXPRESSION ]
);

const OP_DEFINITION = new Recipe(
	"OP_DEFINITION",
	[ 
		new Token("OP_KEYWORD", /^operation/i),
		DEF_SEPARATOR,
		WHITESPACE,
		ASSIGNMENT
	]
);

const ITEMS_DEFINITION = new Recipe(
	"ITEMS_DEFINITION",
	[
		new Token("OP_KEYWORD", /^starting items/i),
		DEF_SEPARATOR,
		WHITESPACE,
		"VALUE_LIST"
	]
);

const LIST_ITEM_SEPARATOR = new Token("LIST_ITEM_SEPARATOR", /^\s*,\s*/, true);

const VALUE_LIST = new Sequence(
	"VALUE_LIST",
	VALUE,
	LIST_ITEM_SEPARATOR
);

const STATEMENT = new Recipe("STATEMENT", [
    new Union("NOISE", [WHITESPACE, EOS]),
    new Union("STATEMENT_BODY", [
        ASSIGNMENT
    ]),
    EOS
]);

const TEST = new Recipe(
	"TEST",
	[
		new Token("TEST_BOILERPLATE", /^Test: divisible by/),
		WHITESPACE,
		NUMBER,
		WHITESPACE,
		new Token("TRUE_OPTION_BOILER", /^If true: throw to monkey/),
		WHITESPACE,
		NUMBER,
		WHITESPACE,
		new Token("FALSE_OPTION_BOILER", /^If false: throw to monkey/),
		WHITESPACE,
		NUMBER
	]
)

const MONKEY_KW = new Token("MONKEY_KW", /^monkey/i);
const MONKEY_ID = new Union("MONKEY_ID", [ INTEGER ]);
const MONKEY = new Recipe(
	"MONKEY",
	[
		MONKEY_KW, WHITESPACE, MONKEY_ID, DEF_SEPARATOR,
			WHITESPACE, ITEMS_DEFINITION,	
			WHITESPACE, OP_DEFINITION,	
			WHITESPACE, TEST,
	]
);

const GAME = new Sequence("GAME", MONKEY, WHITESPACE, false);

export default GAME;
