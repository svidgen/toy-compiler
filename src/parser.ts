export type TreePatternObject = Sequence | Union | Recipe | Token;
export type TreePattern = TreePatternObject | string;

const registry = new Map<string, TreePatternObject>();
const visits = new Map<string, number>();

export class AST {
	public type: string;
	public code: string;
	public start: number;
	public end: number;
	public children: AST[];

	constructor(
		{ type, code, start, end, children }: {
			type: string,
			code: string,
			start: number,
			end: number,
			children: AST[]
		}
	) {
		this.type = type;
		this.code = code;
		this.start = start;
		this.end = end;
		this.children = children;
		// console.log(JSON.stringify(this, null, 2));
		// console.dir(this, { depth: null });
	}

	public get(type: string): AST[] {
		return this.children.filter(c => c.type === type);
	}
}

function codeAt(
	code: string,
	at: number
) {
	let start = at;
	while (start >= 0 && code[start] !== "\n") {
		start--;
	}

	let end = at;
	while (end < code.length && code[end] !== "\n") {
		end++;
	}

	let i = at;
	let lineNumber = 0;
	while (i >= 0) {
		if (code[i] === "\n") lineNumber++;
		i--;
	}

	false && console.log({
		before: code.substring(0, at),
		at,
		after: code.substring(at)
	});

	const lineLabel = `line ${lineNumber}: `;
	const codeline = code.substring(start + 1, end);
	const indicatorIndent = (1 + at - start) + lineLabel.length;
	const indicatorPad = [...new Array(indicatorIndent)].map(i => ' ').join('');
	return `${lineLabel}${codeline}\n${indicatorPad}^`;
}

function raise(
	breadcrumbs: TreePatternObject[],
	nodeType: TreePatternObject,
	code: string,
	at: number,
	message: string
): null {
	const location = codeAt(code, at);
	const trail = breadcrumbs.map(b => b.name).join(':');
	throw new Error(
		`${trail} expected a ${nodeType.name}. ${message}\n${location}`
	);
}

function isRecycle(
	nodeType: TreePatternObject,
	at: number
) {
	const key = `${nodeType.name}:${at}`
	if (!visits.has(key)) {
		visits.set(key, 0);
	}
	visits.set(key, visits.get(key) + 1);
	return visits.get(key) > 32;
}

export class Sequence {
	constructor(
		public name: string,
		public child: TreePattern,
		public delimiter: TreePattern = undefined,
		public optional: boolean = true
	) {
		registry.set(name, this);
	}

	parse({code, at = 0, optional = undefined, breadcrumbs = []}: {code: string, at?: number, optional?: boolean, breadcrumbs?: TreePatternObject[]}): AST | null {
		const allowEmpty = typeof optional === 'boolean' ? optional : this.optional;
		if (isRecycle(this, at)) throw new Error("Recursive token parsing is not allowed!");

		const children: AST[] = [];

		const child = typeof this.child === 'string' ?
			registry.get(this.child) :
			this.child
		;

		const Delimiter = typeof this.delimiter === 'string' ?
			registry.get(this.delimiter) :
			this.delimiter
		;

		let subtree = child.parse({
			code,
			at,
			breadcrumbs: [...breadcrumbs, this],
			optional: allowEmpty
		});

		while (subtree) {
			children.push(subtree);
			let delimiter;
			if (Delimiter) {
				delimiter = Delimiter.parse({
					code,
					at: subtree.end,
					breadcrumbs: [...breadcrumbs, this],
					optional: true
				});
				if (!delimiter) break;
			}
			subtree = child.parse({
				code,
				at: (delimiter || subtree).end,
				breadcrumbs: [...breadcrumbs, this],
				optional: true
			});
		}

		// down the road, this may actually need to return an AST with
		// an empty children collection if allowEmpty is true.
		return children.length > 0 ? new AST({
			type: this.name,
			code: code.substring(at, children[children.length - 1].end),
			start: at,
			end: children[children.length - 1]?.end || at,
			children
		}) : (allowEmpty ? null : raise(
			breadcrumbs, this, code,
			subtree?.end || at,
			`At least one ${child.name} is required.`
		));
	}
}

export class Union {
	constructor(
		public name: string,
		public options: TreePattern[],
		public optional: boolean = false
	) {
		registry.set(name, this);
	}

	parse({code, at = 0, optional = undefined, breadcrumbs = []}: {code: string, at?: number, optional?: boolean, breadcrumbs?: TreePatternObject[]}): AST | null {
		if (isRecycle(this, at)) throw new Error("Recursive token parsing is not allowed!");

		const allowEmpty = typeof optional === 'boolean' ? optional : this.optional;

		for (const option of this.options) {
			const child = typeof option === 'string' ?
				registry.get(option) : option;

			const parsed = child.parse({
				code, at, optional: true,
				breadcrumbs: [...breadcrumbs, this]
			});

			if (parsed) {
				return new AST({
					type: this.name,
					code: code.substring(at, parsed.end),
					start: at,
					end: parsed.end,
					children: [parsed],
				});
			}
		}

		if (allowEmpty) {
			return null;
		} else {
			raise(breadcrumbs, this, code, at, `Must be one of ${
				this.options
					.map(o => typeof o === 'string' ? o : o.name)
					.join(', ')
			}.`);
		}
	}
}

export class Recipe {
	constructor(
		public name: string,
		public children: TreePattern[],
		public optional: boolean = false
	) {
		registry.set(name, this);
	}

	parse({code, at = 0, optional = undefined, breadcrumbs = []}: {code: string, at?: number, optional?: boolean, breadcrumbs?: TreePatternObject[]}): AST | null {
		if (isRecycle(this, at)) throw new Error("Recursive token parsing is not allowed!");

		const children: AST[] = [];
		const allowEmpty = typeof optional === 'boolean' ? optional : this.optional;

		try {
			for (const option of this.children) {
				const child = typeof option === 'string' ?
					registry.get(option) :
					option
				;

				const _at = children.length > 0 ?
					children[children.length - 1].end : at

				const parsed = child.parse({
					code,
					at: _at,
					breadcrumbs: [...breadcrumbs, this],
				});

				if (parsed) {
					children.push(parsed);
				}
			}
		} catch (err) {
			if (allowEmpty) {
				return null;
			} else {
				throw err;
			}
		}

		return new AST({
			type: this.name,
			code: code.substring(at, children[children.length - 1]?.end || at),
			start: at,
			end: children[children.length - 1]?.end || at,
			children,
		});
	}
}

export class Token {
	/**
	 * @param name What to call the token.
	 * @param matcher Whether the given character is part of the value token.
	 */
	constructor(
		public name: string,
		public pattern: RegExp,
		public optional: boolean = false
	) {
		registry.set(name, this);
	}

	parse({code, at = 0, optional = undefined, breadcrumbs = []}: {code: string, at?: number, optional?: boolean, breadcrumbs?: TreePatternObject[]}): AST | null {
		const matched = this.pattern.exec(code.substring(at));
		return (matched && matched.index === 0) ? new AST({
			type: this.name,
			code: matched[0],
			start: at,
			end: at + matched[0].length,
			children: [],
		}) : null;
	}
}
