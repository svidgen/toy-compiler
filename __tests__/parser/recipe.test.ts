import { recipe, sequence, token, union } from "../../src/parser";

describe("recipe", () => {
	const LETTER_A = token("LETTER_A", /[aA]/);
	const LETTER_B = token("LETTER_B", /[bB]/);
	const LETTER_C = token("LETTER_C", /[cC]/);
	const RECIPE_ABC = recipe("RECIPE_ABC", [LETTER_A, LETTER_B, LETTER_C]);

	it("can parse a simple, static recipe", () => {
		const code = "abc";
		const ast = RECIPE_ABC.parse({ code });
		expect(ast).toMatchSnapshot();
	});

	/**
	 * TODO: I think these are wrong ... they shouldn't match. They should throw
	 * exceptions telling me what tokens are expected.
	 *
	 * Maybe we fix this more broadly by changing the way we handle optional tokens.
	 *
	 * Instead of trying to handle optionality at every language structure, we create
	 * an `optional()` wrapper that swallows errors when they're thrown. This doesn't
	 * seem super efficient though.
	 *
	 * Alternatively, perhaps recipes need to inspect the optionality of their
	 * children when nulls come back from child parses. TBD.
	 */
	for (const code of ["a", "b", "c", "bac", "cab", "cba", "ab", "bc"]) {
		it(`fails on non-matching string '${code}'`, () => {
			const ast = RECIPE_ABC.parse({ code });
			console.log({ code, ast });
		});
	}
});
