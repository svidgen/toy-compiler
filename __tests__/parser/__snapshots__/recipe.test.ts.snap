// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`recipe can parse a simple, static recipe 1`] = `
AST {
  "children": [
    AST {
      "children": [],
      "code": "a",
      "end": 1,
      "start": 0,
      "type": "LETTER_A",
    },
    AST {
      "children": [],
      "code": "b",
      "end": 2,
      "start": 1,
      "type": "LETTER_B",
    },
    AST {
      "children": [],
      "code": "c",
      "end": 3,
      "start": 2,
      "type": "LETTER_C",
    },
  ],
  "code": "abc",
  "end": 3,
  "start": 0,
  "type": "RECIPE_ABC",
}
`;
